Create github-vs-sourcegraph.md (#44047)
sourcegraph/sourcegraph: doc/getting-started/github-vs-sourcegraph.md

**Forks** are included in the index, but they are subject to the same limitations as other repositories, so not all forks are indexed. You may need to include the fork filter to retrieve results for the fork repos, but an admin can adjust global settings to automatically include forks in search query results. 

In addition to searching your organization’s private code, you can use Sourcegraph.com to search across 2.8 million public repositories from multiple code hosts. 

<table>
  <tr>
   <td>
   </td>
  <td><strong>GitHub</strong>
   </td>
   <td><strong>Sourcegraph</strong>

   </td>
  </tr>
  <tr>
   <td><strong>Search across all repositories and forks</strong>

   </td>
   <td>✓ with limitations 

   </td>
   <td>✓ with limitations 

   </td>
  </tr>
  <tr>
   <td><strong>Search across files larger than 350 KiB</strong>

   </td>
   <td>✗

   </td>
   <td>✓ 

Using the [search.largeFiles](https://docs.sourcegraph.com/admin/config/site_config#search-largeFile) keyword

   </td>
  </tr>
  <tr>
   <td><strong>Search across all branches</strong>

   </td>
   <td>Only the default branch is searchable

   </td>
   <td>✓

   </td>
  </tr>
</table>

### Search syntax

> Sourcegraph offers [structural search](https://docs.sourcegraph.com/code_search/reference/structural), and GitHub code search does not offer this search method. Structural search lets you match richer syntax patterns, specifically in code and structured data formats like JSON. Sourcegraph offers structural search on indexed code and uses [Comby syntax](https://comby.dev/docs/syntax-reference) for structural matching of code blocks or nested expressions. For example, the fmt.Sprintf function is a popular print function in Go. [Here](https://sourcegraph.com/search?q=repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+fmt.Sprintf%28...%29&patternType=structural&_ga=2.204781593.827352295.1667227568-1057140468.1661198534&_gac=1.118615675.1665776224.CjwKCAjwkaSaBhA4EiwALBgQaJCOc6GlhIDQyg6HQScgfSBQpoFTUf7T_NNqEX5JaobtCS08GUEJuRoCIlIQAvD_BwE&_gl=1*1r2u5zs*_ga*MTA1NzE0MDQ2OC4xNjYxMTk4NTM0*_ga_E82CCDYYS1*MTY2NzUwODExNC4xMTQuMS4xNjY3NTA5NjUyLjAuMC4w) is a pattern that matches all of the arguments in fmt.Sprintf in our code using structural search compared to the [search](https://sourcegraph.com/search?q=context:global+repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+fmt.Sprintf%28...%29&patternType=regexp) using regex. 

Reviewer: @lcolston08 - update the table so "Structural search" is first.

Author: Updated

<table>
  <tr>
   <td>
   </td>
   <td><strong>GitHub</strong>
   </td>
   <td><strong>Sourcegraph</strong>
   </td>
  </tr>
  <tr>
   <td><strong>Literal search </strong>
   </td>
   <td> ✓
   </td>
   <td>✓
   </td>
  </tr>
  <tr>
   <td><strong>Regular expression search</strong>
   </td>
   <td> ✓
   </td>
   <td>✓
   </td>
  </tr>
  <tr>
   <td><strong>Structural search </strong>
   </td>
   <td>✗
   </td>
   <td>✓
   </td>
  </tr>
  <tr>
   <td><strong>Multi-line search support</strong>
   </td>
   <td>✗
   </td>
   <td>✓
   </td>
  </tr>
  <tr>
   <td><strong>Commit diff and commit message searches</strong>
   </td>
   <td>✗
   </td>
   <td>✓
   </td>
  </tr>
  <tr>
   <td><strong>Symbol search</strong>
   </td>
   <td>10 languages
<p>
C#, Python, Go, Java, JavaScript, TypeScript, PHP, Protocol Buffers, Ruby, and Rust.
   </td>
   <td>75+ languages
   </td>
  </tr>
</table>

### Search results and result types

**Search results**

GitHub only returns the first 10 pages of search results. You cannot currently go past the 10th page or retrieve all search results.

> Sourcegraph can retrieve all search results. By default, Sourcegraph returns 500 search results, but this number can be increased by increasing the ‘count’ value. Sourcegraph can display a maximum of 1500 matches, but all matches can be fetched using the [src CLI](https://docs.sourcegraph.com/cli/quickstart), the [Stream API](https://docs.sourcegraph.com/api/stream_api), or [GraphQL API](https://docs.sourcegraph.com/api/graphql). You can also export the results via CSV. 

Reviewer: ```suggestion
Sourcegraph can retrieve all search results. By default, Sourcegraph returns 500 search results, but this number can be increased by increasing the ‘count’ value. Sourcegraph can display a maximum of 1,500 matches, but all matches can be fetched using the [src CLI](https://docs.sourcegraph.com/cli/quickstart), the [Stream API](https://docs.sourcegraph.com/api/stream_api), or [GraphQL API](https://docs.sourcegraph.com/api/graphql). You can also export the results via CSV. 
```

* Up rank files with a lot of symbols
* Up rank small files
* Up rank short names
* Up rank branch count
* Down rank generated code
* Down rank vendored code
* Down rank test code

When submitting a search query, the quality of a match is scored based on language-specific heuristics (Java classes rank higher than variables), word boundaries, and symbol ranges. The score that is received at the time of the query is combined with the index time rank and the repository’s priority to determine the final ranking.

<table>
  <tr>
   <td>
   </td>
   <td><strong>GitHub</strong>
   </td>
   <td><strong>Sourcegraph</strong>
   </td>
  </tr>
  <tr>
   <td><strong>Comprehensive search results</strong>
   </td>
   <td>✗
<p>
Limited to 10 pages of results
   </td>
   <td>✓ 
<p>

> 500 searches are returned by default, but this number can be increased by increasing the ‘count’ value. A maximum of 1500 matches can be displayed, and more matches can be fetched using the src CLI, the Stream API, or GraphQL API.

Reviewer: ```suggestion
500 searches are returned by default, but this number can be increased by increasing the ‘count’ value. A maximum of 1,500 matches can be displayed, and more matches can be fetched using the src CLI, the Stream API, or GraphQL API.
```

* “Find implementations,” which allows you to navigate to a symbol’s interface definition or find all the places an interface is being implemented.

Sourcegraph’s precise code navigation is opt-in and requires you to upload code graph data ([LSIF or SCIP](https://docs.sourcegraph.com/code_navigation/references/indexers)) to Sourcegraph. This data can be automatically generated and uploaded to Sourcegraph via [auto-indexing](https://docs.sourcegraph.com/code_navigation/explanations/auto_indexing). For repositories without SCIP or LSIF data, Sourcegraph automatically falls back to search-based code navigation.

Sourcegraph’s precise code navigation [supports 11 languages](https://docs.sourcegraph.com/code_navigation/references/indexers). 

<table>
  <tr>
   <td>
   </td>
   <td><strong>GitHub</strong>
   </td>
   <td><strong>Sourcegraph</strong>
   </td>
  </tr>
  <tr>
   <td><strong>Technical implementation</strong>
   </td>
   <td>Heuristic-based
   </td>
   <td>SCIP-based (code graph data)
   </td>
  </tr>
  <tr>
   <td><strong>Accuracy</strong>
   </td>
   <td>High (some false positives)
   </td>

>    <td>Perfect (compiler accurate)

Reviewer: Should this be compiler-accurate

</li>
</ul>
   </td>
  </tr>
  <tr>
   <td><strong>Search across all branches</strong>
   </td>
   <td>Only the default branch is searchable
   </td>
   <td>✓
   </td>
  </tr>
  <tr>
   <td><strong>Search across all forks </strong>
   </td>
   <td>✓ 
<p>
Limitations include: 
<ul>

<li>Files over 350 KiB and empty files are excluded

<li>Only UTF-8 encoded files are included

<li>Very large repositories may not be indexed

<li>Vendored and generated code is excluded (as determined by <a href="https://github.com/go-enry/go-enry">Enry</a>)
</li>
</ul>
   </td>
   <td>✓
<p>
Sourcegraph allows you to search indexed and unindexed code within forks. 
<p>
Limitations for code that is indexed: 
<ul>

<li>Files larger than 1 MB*

<li>Binary files

<li>Files other than UTF-8 

<p>
*You can use the <a href="https://docs.sourcegraph.com/admin/config/site_config#search-largeFile">search.largeFiles</a> keyword to specify files to be indexed and searched regardless of size.
</li>
</ul>
   </td>
  </tr>
</table>

### Search syntax

> Both GitHub code search and Sourcegraph support regular expression and literal search. [Regular expression](https://docs.sourcegraph.com/code_search/reference/queries#standard-search-default) helps you find code that matches a pattern (including classes of characters like letters, numbers, and whitespace) and can restrict the results to anchors like the start of a line, the end of a line, or word boundary. Literal (standard) search matches literal patterns exactly, including punctuation, like quotes.

Author: ```suggestion
Sourcegraph offers [structural search](https://docs.sourcegraph.com/code_search/reference/structural), and GitHub code search does not offer this search method. Structural search lets you match richer syntax patterns, specifically in code and structured data formats like JSON. Sourcegraph offers structural search on indexed code and uses [Comby syntax](https://comby.dev/docs/syntax-reference) for structural matching of code blocks or nested expressions. For example, the fmt.Sprintf function is a popular print function in Go. [Here](https://sourcegraph.com/search?q=repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+fmt.Sprintf%28...%29&patternType=structural&_ga=2.204781593.827352295.1667227568-1057140468.1661198534&_gac=1.118615675.1665776224.CjwKCAjwkaSaBhA4EiwALBgQaJCOc6GlhIDQyg6HQScgfSBQpoFTUf7T_NNqEX5JaobtCS08GUEJuRoCIlIQAvD_BwE&_gl=1*1r2u5zs*_ga*MTA1NzE0MDQ2OC4xNjYxMTk4NTM0*_ga_E82CCDYYS1*MTY2NzUwODExNC4xMTQuMS4xNjY3NTA5NjUyLjAuMC4w) is a pattern that matches all of the arguments in fmt.Sprintf in our code using structural search compared to the [search](https://sourcegraph.com/search?q=context:global+repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+fmt.Sprintf%28...%29&patternType=regexp) using regex. 

Both GitHub code search and Sourcegraph support regular expression and literal search. [Regular expression](https://docs.sourcegraph.com/code_search/reference/queries#standard-search-default) helps you find code that matches a pattern (including classes of characters like letters, numbers, and whitespace) and can restrict the results to anchors like the start of a line, the end of a line, or word boundary. Literal (standard) search matches literal patterns exactly, including punctuation, like quotes.
```

<table>
  <tr>
  </tr>
</table>

GitHub’s search syntax can be found [here](https://cs.github.com/about/syntax), and Sourcegraph’s search syntax can be found [here](https://docs.sourcegraph.com/code_search/reference/queries). 

> In addition to regular expression and literal search, Sourcegraph offers [structural search.](https://docs.sourcegraph.com/code_search/reference/structural) GitHub code search does not offer this search method. Structural search lets you match richer syntax patterns, specifically in code and structured data formats like JSON. Sourcegraph offers structural search on indexed code and uses [Comby syntax](https://comby.dev/docs/syntax-reference) for structural matching of code blocks or nested expressions. For example, the fmt.Sprintf function is a popular print function in Go. [Here](https://sourcegraph.com/search?q=repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+fmt.Sprintf%28...%29&patternType=structural&_ga=2.204781593.827352295.1667227568-1057140468.1661198534&_gac=1.118615675.1665776224.CjwKCAjwkaSaBhA4EiwALBgQaJCOc6GlhIDQyg6HQScgfSBQpoFTUf7T_NNqEX5JaobtCS08GUEJuRoCIlIQAvD_BwE&_gl=1*1r2u5zs*_ga*MTA1NzE0MDQ2OC4xNjYxMTk4NTM0*_ga_E82CCDYYS1*MTY2NzUwODExNC4xMTQuMS4xNjY3NTA5NjUyLjAuMC4w) is a pattern that matches all of the arguments in fmt.Sprintf in our code using structural search compared to the [search](https://sourcegraph.com/search?q=context:global+repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+fmt.Sprintf%28...%29&patternType=regexp) using regex. 

Author: ```suggestion
Remove this section:
In addition to regular expression and literal search, Sourcegraph offers [structural search.](https://docs.sourcegraph.com/code_search/reference/structural) GitHub code search does not offer this search method. Structural search lets you match richer syntax patterns, specifically in code and structured data formats like JSON. Sourcegraph offers structural search on indexed code and uses [Comby syntax](https://comby.dev/docs/syntax-reference) for structural matching of code blocks or nested expressions. For example, the fmt.Sprintf function is a popular print function in Go. [Here](https://sourcegraph.com/search?q=repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+fmt.Sprintf%28...%29&patternType=structural&_ga=2.204781593.827352295.1667227568-1057140468.1661198534&_gac=1.118615675.1665776224.CjwKCAjwkaSaBhA4EiwALBgQaJCOc6GlhIDQyg6HQScgfSBQpoFTUf7T_NNqEX5JaobtCS08GUEJuRoCIlIQAvD_BwE&_gl=1*1r2u5zs*_ga*MTA1NzE0MDQ2OC4xNjYxMTk4NTM0*_ga_E82CCDYYS1*MTY2NzUwODExNC4xMTQuMS4xNjY3NTA5NjUyLjAuMC4w) is a pattern that matches all of the arguments in fmt.Sprintf in our code using structural search compared to the [search](https://sourcegraph.com/search?q=context:global+repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+fmt.Sprintf%28...%29&patternType=regexp) using regex. 
```

## Searching code

### Code host integrations

GitHub code search can only be used to search across code that is stored in GitHub. Organizations with code stored in multiple code hosts cannot use GitHub code search to search across their entire codebase. 

Sourcegraph integrates with multiple code hosts to provide universal code search across all of your organization’s code, no matter where it’s stored. Sourcegraph has out-of-the-box [integrations](https://docs.sourcegraph.com/integration) with: 

* GitHub
* GitLab
* Bitbucket Cloud and Bitbucket Server / Bitbucket Data Center
* Perforce

You can also integrate Sourcegraph with other Git-based code hosts using [these](https://docs.sourcegraph.com/admin/external_service/other) instructions or use the [Sourcegraph CLI (src)](https://docs.sourcegraph.com/admin/external_service/src_serve_git) to load local repositories without a code host into Sourcegraph. Sourcegraph is continuously adding [Tier 1](https://docs.sourcegraph.com/admin/external_service) support for additional code hosts. 

> **Code host integrations**

Reviewer 2: This might be too pedantic, but "integration" conveys a certain level of support. In my mind, code host integration means that it appears as an option within Sourcegraph to sync code from. This is true of the named code hosts in the table below. 

If a user has to use the Sourcegraph CLI (which is required for the last row: Any git-based code host), then I'd argue we don't have an "integration" with that code host. Instead, we support syncing code from that code host even though it's not actually integrated. I think a more correct name for this table is "Where can [tool] sync code from". I know that's not as clean of a title :)

Reviewer: That's a good point. We could probably remove the title from the table (none of the other tables have titles).

Author: The title has been removed. Closing this out.

> <table>

Reviewer 2: Honestly I find this table kind of confusing, especially because (I think) it's just rehashing what the previous paragraphs are saying. The fact that the table contains bullets and checks and normal text and asterisks made it hard to parse. 

I wonder if individual sub-sections (Branches, Forks, Repositories) with a tight blurb underneath each comparing GH and Sourcegraph would be cleaner.

Author: I see what you're both saying, and it's valid feedback. We'd have to reorder the section to focus on the three topics in order and alternate between SG and GH.

Repos

- GH
- SG
- Chart

Branches
- GH
- SG
- Chart

Forks
- GH
- SG
- Chart

I'll propose the changes in the word doc @abschu

Author: GitHub allows you to search indexed **repositories**, but not all code is indexed. GitHub’s [current limitations](https://cs.github.com/about/faq#indexed-content) on indexed code are:

* Files over 350 KiB and empty files are excluded
* Only UTF-8 encoded files are included
* Very large repositories may not be indexed
* Vendored and generated code is excluded (as determined by [Enry](https://github.com/go-enry/go-enry))

Sourcegraph allows you to search indexed and [unindexed](https://docs.sourcegraph.com/code_search/how-to/exhaustive#non-indexed-backends) code. Sourcegraph’s [current limitations](https://docs.sourcegraph.com/admin/search) on indexed code are: 

* Files larger than 1 MB are excluded* 
* Binary files are excluded 
* Files other then UTF-9 are excluded 

*You can use the [search.largeFiles](https://docs.sourcegraph.com/admin/config/site_config#search-largeFile) keyword to specify files to be indexed and searched regardless of size.

<table>
  <tr>
   <td>
   </td>
   <td><strong>GitHub</strong>
   </td>
   <td><strong>Sourcegraph</strong>
   </td>
  </tr>
  <tr>
   <td><strong>Search across all repositories</strong>
   </td>
   <td>Limitations include: 
<ul>

<li>Files over 350 KiB and empty files are excluded

<li>Only UTF-8 encoded files are included

<li>Very large repositories may not be indexed

<li>Vendored and generated code is excluded (as determined by <a href="https://github.com/go-enry/go-enry">Enry</a>)
</li>
</ul>
   </td>
   <td>Sourcegraph allows you to search indexed and unindexed code. 
<p>
Limitations for code that is indexed: 
<ul>

<li>Files larger than 1 MB*

<li>Binary files

<li>Files other than UTF-8 

<p>
*You can use the <a href="https://docs.sourcegraph.com/admin/config/site_config#search-largeFile">search.largeFiles</a> keyword to specify files to be indexed and searched regardless of size.
</li>
</ul>
   </td>
  </tr>
</table>

With GitHub, only the default **branch** is searchable (though GitHub is planning to support branch search in the future). 

With Sourcegraph, typically, the latest code on the default **branch** of each repository is indexed (usually the master or main), but Sourcegraph can also index other non-default branches, such as long-running branches like release branches. If you’re searching outside of indexed branches, you can use unindexed search. You should expect slightly slower results when searching unindexed code.  

<table>
  <tr>
   <td>
   </td>
   <td><strong>GitHub</strong>
   </td>
   <td><strong>Sourcegraph</strong>
   </td>
  </tr>
  <tr>
   <td><strong>Search across all branches</strong>
   </td>
   <td>Only the default branch is searchable
   </td>
   <td>✓
   </td>
  </tr>
</table>

With GitHub, **forks** are included in the index, but they are subject to the same limitations as other repositories, so not all forks are indexed.

With Sourcegraph, **Forks** are included in the index, but they are subject to the same limitations as other repositories, so not all forks are indexed. You may need to include the fork filter to retrieve results for the fork repos, but an admin can adjust global settings to automatically include forks in search query results. 

<table>
  <tr>
   <td>
   </td>
   <td><strong>GitHub</strong>
   </td>
   <td><strong>Sourcegraph</strong>
   </td>
  </tr>
  <tr>
   <td><strong>Search across all forks </strong>
   </td>
   <td>Limitations include: 
<ul>

<li>Files over 350 KiB and empty files are excluded

<li>Only UTF-8 encoded files are included

<li>Very large repositories may not be indexed

<li>Vendored and generated code is excluded (as determined by <a href="https://github.com/go-enry/go-enry">Enry</a>)
</li>
</ul>
   </td>
   <td>Sourcegraph allows you to search indexed and unindexed code within forks. 
<p>
Limitations for code that is indexed: 
<ul>

<li>Files larger than 1 MB*

<li>Binary files

<li>Files other than UTF-8 

<p>
*You can use the <a href="https://docs.sourcegraph.com/admin/config/site_config#search-largeFile">search.largeFiles</a> keyword to specify files to be indexed and searched regardless of size.
</li>
</ul>
   </td>
  </tr>
</table>

In addition to searching your private code, GitHub has indexed over 7 million public GitHub repositories which are also searchable.

With Sourcegraph, in addition to searching your organization’s private code, you can use Sourcegraph.com to search across 2.8 million public repositories from multiple code hosts.

Author: Spoke with Andy and we're going to make the chart easier to follow and keep it where it is with the current flow of content. 

  <tr>
   <td>
   </td>
   <td>**GitHub**

   </td>
   <td>**Sourcegraph**

   </td>
  </tr>
  <tr>
   <td>**Search across all repositories and forks**

   </td>
   <td>✓ with limitations 

   </td>
   <td>✓ with limitations 

   </td>
  </tr>
  <tr>
   <td>**Sarch across files larger than 350 KiB**

   </td>
   <td>✗

   </td>
   <td>✓ 

Using the [search.largeFiles](https://docs.sourcegraph.com/admin/config/site_config#search-largeFile) keyword

   </td>
  </tr>
  <tr>
   <td>**Search across all branches**

   </td>
   <td>Only the default branch is searchable

   </td>
   <td>✓

   </td>
  </tr>

Author: Resolving since we updated the table.

</table>

### Searching repositories, branches, and forks

GitHub allows you to search indexed **repositories**, but not all code is indexed. GitHub’s [current limitations](https://cs.github.com/about/faq#indexed-content) on indexed code are:

* Files over 350 KiB and empty files are excluded
* Only UTF-8 encoded files are included
* Very large repositories may not be indexed
* Vendored and generated code is excluded (as determined by [Enry](https://github.com/go-enry/go-enry))

With GitHub, only the default **branch** is searchable (though GitHub is planning to support branch search in the future). 

**Forks** are included in the index, but they are subject to the same limitations as other repositories, so not all forks are indexed.

In addition to searching your private code, GitHub has indexed over 7 million public GitHub repositories which are also searchable.

Sourcegraph allows you to search indexed and [unindexed](https://docs.sourcegraph.com/code_search/how-to/exhaustive#non-indexed-backends) code. Sourcegraph’s [current limitations](https://docs.sourcegraph.com/admin/search) on indexed code are: 

* Files larger than 1 MB are excluded* 
* Binary files are excluded 

> * Files other then UTF-9 are excluded 

Reviewer 2: ```suggestion
* Files other than UTF-8 are excluded 
```

*You can use the [search.largeFiles](https://docs.sourcegraph.com/admin/config/site_config#search-largeFile) keyword to specify files to be indexed and searched regardless of size.

With Sourcegraph, typically, the latest code on the default **branch** of each repository is indexed (usually the master or main), but Sourcegraph can also index other non-default branches, such as long-running branches like release branches. If you’re searching outside of indexed branches, you can use unindexed search. You should expect slightly slower results when searching unindexed code.  

**Forks** are included in the index, but they are subject to the same limitations as other repositories, so not all forks are indexed. You may need to include the fork filter to retrieve results for the fork repos, but an admin can adjust global settings to automatically include forks in search query results. 

> In addition to searching your organization’s private code, you can use Sourcegraph.com to search across 2.8 million public repositories from multiple code hosts. 

Reviewer 2: Are we indexing code from code hosts other than GitHub on dot com?

Reviewer 3: Yes, for example we have repos from http://src.fedoraproject.org/

Author: Thanks. This seems to back up the statement about public repos from multiple code hosts.

</li>
</ul>
   </td>
  </tr>
  <tr>
   <td><strong>Search across all branches</strong>
   </td>
   <td>Only the default branch is searchable
   </td>
   <td>✓
   </td>
  </tr>
  <tr>
   <td><strong>Search across all forks </strong>
   </td>
   <td>✓ 
<p>
Limitations include: 
<ul>

<li>Files over 350 KiB and empty files are excluded

<li>Only UTF-8 encoded files are included

<li>Very large repositories may not be indexed

<li>Vendored and generated code is excluded (as determined by <a href="https://github.com/go-enry/go-enry">Enry</a>)
</li>
</ul>
   </td>
   <td>✓
<p>
Sourcegraph allows you to search indexed and unindexed code within forks. 
<p>
Limitations for code that is indexed: 
<ul>

<li>Files larger than 1 MB*

<li>Binary files

<li>Files other than UTF-8 

<p>
*You can use the <a href="https://docs.sourcegraph.com/admin/config/site_config#search-largeFile">search.largeFiles</a> keyword to specify files to be indexed and searched regardless of size.
</li>
</ul>
   </td>
  </tr>
</table>

> ### Search syntax

Author: I tried to reorder this and made proposals below

### Commit diff search and commit message search

Commit diff and commit message searches help you see how your codebase has changed over time. With Sourcegraph, you can search within [commit diffs](https://docs.sourcegraph.com/code_search/explanations/features#commit-diff-search) to find changes to particular functions, classes, or specific areas of the codebase, and you can search over [commit messages](https://docs.sourcegraph.com/code_search/explanations/features#commit-message-search) and narrow down searches with additional filters such as author or time. 

> Neither commit diff search nor commit message search is available with GitHub code search. 

Reviewer 2: ```suggestion
GitHub code search does not offer commit diff search or commit message search. 
```

<table>
  <tr>
   <td>
   </td>
   <td><strong>GitHub</strong>
   </td>
   <td><strong>Sourcegraph</strong>
   </td>
  </tr>
  <tr>
   <td><strong>Literal search </strong>
   </td>
   <td> ✓
   </td>
   <td>✓
   </td>
  </tr>
  <tr>
   <td><strong>Regular expression search</strong>
   </td>
   <td> ✓
   </td>
   <td>✓
   </td>
  </tr>
  <tr>
   <td><strong>Structural search </strong>
   </td>
   <td>✗
   </td>
   <td>✓
   </td>
  </tr>
  <tr>
   <td><strong>Multi-line search support</strong>
   </td>
   <td>✗
   </td>
   <td>✓
   </td>
  </tr>
  <tr>
   <td><strong>Commit diff and commit message searches</strong>
   </td>
   <td>✗
   </td>
   <td>✓
   </td>
  </tr>
  <tr>
   <td><strong>Symbol search</strong>
   </td>
   <td>10 languages
<p>
C#, Python, Go, Java, JavaScript, TypeScript, PHP, Protocol Buffers, Ruby, and Rust.
   </td>
   <td>75+ languages
   </td>
  </tr>
</table>

### Search results and result types

**Search results**

GitHub only returns the first 10 pages of search results. You cannot currently go past the 10th page or retrieve all search results.

> All search results can be retrieved with Sourcegraph. By default, 500 search results are returned, but this number can be increased by increasing the ‘count’ value. A maximum of 1500 matches can be displayed, but all matches can be fetched using the [src CLI](https://docs.sourcegraph.com/cli/quickstart), the [Stream API](https://docs.sourcegraph.com/api/stream_api), or [GraphQL API](https://docs.sourcegraph.com/api/graphql). You can also export the results via CSV. 

Reviewer 2: ```suggestion
Sourcegraph can retrieve all search results. By default, 500 search results are returned, but this number can be increased by increasing the ‘count’ value. A maximum of 1500 matches can be displayed, but all matches can be fetched using the [src CLI](https://docs.sourcegraph.com/cli/quickstart), the [Stream API](https://docs.sourcegraph.com/api/stream_api), or [GraphQL API](https://docs.sourcegraph.com/api/graphql). You can also export the results via CSV. 
```

<table>
  <tr>
   <td>
   </td>
   <td><strong>GitHub</strong>
   </td>
   <td><strong>Sourcegraph</strong>
   </td>
  </tr>
  <tr>
   <td><strong>Filters</strong>
   </td>
   <td>✓ 
   </td>
   <td>✓ 
   </td>
  </tr>
  <tr>
   <td><strong>Search contexts</strong>
   </td>
   <td>✓ 
<p>
Simple and advanced
   </td>
   <td>✓ 
<p>
Repository-based and query-based
   </td>
  </tr>
</table>

## Understanding and navigating code

**Code navigation** helps you explore code in depth. It includes features such as “Go to definition” and “Find references,” which let you quickly move between files to understand code.

### Search-based code navigation

Both GitHub and Sourcegraph offer out-of-the-box, search-based code navigation. This version of code navigation uses search-based heuristics to find references and definitions across a codebase without any setup required. It is powerful and convenient, but it can sometimes present inaccurate results. For example, it can return false positive references for symbols with common names. It is limited to returning definitions and references within a single repository (it won’t track references across multiple repositories).

> GitHub’s [search-based code navigation](https://docs.github.com/en/repositories/working-with-files/using-files/navigating-code-on-github) supports 10 languages. Sourcegraph’s [search-based code navigation](https://docs.sourcegraph.com/code_navigation/explanations/search_based_code_navigation) supports 40 languages. Sourcegraph also uses a technology called Squirrel for Java and Python, providing higher accuracy for search-based code navigation for those languages.

Reviewer 2: Mentioning "a technology called Squirrel" strikes me as an unncessary implementation detail, especially because we don't quantify how beneficial it is.

Reviewer: CC @iskyOS

Reviewer 3: That's included to justify why some of our languages have "high" accuracy, but we don't have a docs page unpacking it, otherwise I'd link out. We can strike that sentence since it doesn't add much right now.

```suggestion
GitHub’s [search-based code navigation](https://docs.github.com/en/repositories/working-with-files/using-files/navigating-code-on-github) supports 10 languages. Sourcegraph’s [search-based code navigation](https://docs.sourcegraph.com/code_navigation/explanations/search_based_code_navigation) supports 40 languages.
```

Author: I think this is the right call given Jordan's feedback. I'm going to commit the suggestion to remove the last sentence.

<table>
  <tr>
   <td>
   </td>
   <td><strong>GitHub</strong>
   </td>
   <td><strong>Sourcegraph</strong>
   </td>
  </tr>
  <tr>
   <td><strong>Technical implementation</strong>
   </td>
   <td>Heuristic-based
   </td>
   <td>Heuristic-based
   </td>
  </tr>
  <tr>
   <td><strong>Language support</strong>
   </td>
   <td>10 languages
<p>
C#, CodeQL, Elixir, Go, Java, JavaScript, TypeScript, PHP, Python, Ruby
   </td>
   <td>40 languages
<p>
<a href="https://docs.sourcegraph.com/code_navigation/explanations/search_based_code_navigation">The full list of supported languages can be found here. </a>
   </td>
  </tr>
  <tr>
   <td><strong>Setup</strong>
   </td>
   <td>No setup required
   </td>
   <td>No setup required
   </td>
  </tr>
  <tr>
   <td><strong>Accuracy</strong>
   </td>
   <td>Moderate (some false positives)
   </td>

>    <td>Moderate to high based on language (some false positives)

Reviewer 2: If I'm a discerning reader, I'm going to lose some trust if I see a row titled "Accuracy" but we're using imprecise language like "moderate" and "moderate to high". I'd expect to see data.

Reviewer: CC @iskyOS

Reviewer 3: Unfortunately we don't have a quantifiable way to describe the accuracy differences. From my understanding, accuracy is variable based on if a codebase is using common names for symbols (because that's where false positives pop up).

We can change this to "Some false positives" for both to be more objective.

Reviewer 3: Let's remove the "moderate to high" wording, because I think that makes this feel extra imprecise. I think "moderate" is OK for now until we have real data or a metric.

```suggestion
   <td>Moderate (some false positives)
```

Reviewer 3: @lcolston08

   </td>
  </tr>
  <tr>
   <td><strong>Cross-repository</strong>
   </td>
   <td>✗
   </td>
   <td>✗
   </td>
  </tr>
</table>

### Precise code navigation

> GitHub and Sourcegraph both offer precise code navigation as well. Despite having the same name, the two versions of precise code navigation are very different in terms of the underlying technology and accuracy they provide. Only Sourcegraph’s precise code navigation is 100% accurate. 

Reviewer 2: We're 100% sure that it's 100% accurate?

Reviewer: cc @iskyOS

Reviewer 3: It's always accurate so long as the code graph data is fresh. Code graph data can get stale once code has been changed, but in that case, I believe we're falling back to search-based navigation anyway.

@macraig Is that understanding correct?

Reviewer 4: @iskyOS correct, we can also say Only SG's precise code nav is compiler-accurate or IDE-accurate if saying 100% sounds too sales-y.

Reviewer 3: I don't personally think it's salesy; I would leave as-is given it sounds like a bold (and true) statement.

Reviewer 3: FYI @lcolston08

Author: Thanks. Resolving this conversation.

* Cross-repository navigation, which shows symbol usage across repositories and [transitive dependencies](https://docs.sourcegraph.com/code_navigation/explanations/features#beta-dependency-navigation). Dependency navigation is currently in beta.
* “Find implementations,” which allows you to navigate to a symbol’s interface definition or find all the places an interface is being implemented.

Sourcegraph’s precise code navigation is opt-in and requires you to upload code graph data ([LSIF or SCIP](https://docs.sourcegraph.com/code_navigation/references/indexers)) to Sourcegraph. This data can be automatically generated and uploaded to Sourcegraph via [auto-indexing](https://docs.sourcegraph.com/code_navigation/explanations/auto_indexing). For repositories without SCIP or LSIF data, Sourcegraph automatically falls back to search-based code navigation.

Sourcegraph’s precise code navigation [supports 11 languages](https://docs.sourcegraph.com/code_navigation/references/indexers). 

<table>
  <tr>
   <td>
   </td>
   <td><strong>GitHub</strong>
   </td>
   <td><strong>Sourcegraph</strong>
   </td>
  </tr>
  <tr>

>    <td><strong>Technical implementation</strong>

Reviewer 2: Not sure if this is a necessary row for the end user

Reviewer: cc @iskyOS

Reviewer 3: I agree that the detail itself isn't that interesting, but until we have a better way to visualize heuristic-based vs code graph-based code nav, I think we should keep this in.

### Codebase insights

GitHub does not offer comprehensive insights that account for the content of the code itself. Rather, GitHub’s insights are based primarily on GitHub’s product-level data. GitHub offers [dependency insights](https://docs.github.com/en/enterprise-cloud@latest/admin/policies/enforcing-policies-for-your-enterprise/enforcing-policies-for-dependency-insights-in-your-enterprise) that show all the packages your organization’s repositories depend on, e.g. aggregated information about security advisories and licenses. 

[Sourcegraph’s Code Insights](https://docs.sourcegraph.com/code_insights) is based on codebase-level data: aggregation of lines, patterns, and other search targets in the codebase. It reveals high-level information about the entire codebase, accounting for all of your repositories and code hosts. With Code Insights, you can track anything that can be expressed with a Sourcegraph search query and turn it into customizable dashboards. Code Insights can be used to track migrations, package use, version adoption, code smells, vulnerability remediation, codebase size, and more. 

###  App activity insights

> With [insights for Projects](https://docs.github.com/en/issues/planning-and-tracking-with-projects/viewing-insights-from-your-project/about-insights-for-projects), you can view, create, and customize charts that are built from the project’s data. [Organization activity insights](https://docs.github.com/en/enterprise-cloud@latest/organizations/collaborating-with-groups-in-organizations/viewing-insights-for-your-organization) help you understand how members of the organization are using GitHub, e.g. issue and pull request activity, top languages used, and cumulative information about where members spend their time. 

Reviewer 2: ```suggestion
With [insights for Projects](https://docs.github.com/en/issues/planning-and-tracking-with-projects/viewing-insights-from-your-project/about-insights-for-projects) in GitHub, you can view, create, and customize charts that are built from the project’s data. [Organization activity insights](https://docs.github.com/en/enterprise-cloud@latest/organizations/collaborating-with-groups-in-organizations/viewing-insights-for-your-organization) help you understand how members of the organization are using GitHub, e.g. issue and pull request activity, top languages used, and cumulative information about where members spend their time. 
```

Sourcegraph’s [editor integrations](https://docs.sourcegraph.com/integration/editor) let you search and navigate across all of your repositories from all your code hosts and across all GitHub instances and organizations without leaving your IDE. Sourcegraph currently integrates with VS Code, JetBrains IDEs, Sublime Text, Atom, and Gitpod. You can also add Sourcegraph to your preferred [browser](https://docs.sourcegraph.com/integration/browser_extension/how-tos/browser_search_engine) to quickly search across your entire codebase from within your browser. 

### API

> GitHub has a REST API for web clients, but it is not yet documented. On the other hand, Sourcegraph offers a few different APIs that help you access data related to the code available on a Sourcegraph instance. The [GraphQL API](https://docs.sourcegraph.com/api/graphql) is for accessing data stored and computed by Sourcegraph. This API can be [used](https://docs.sourcegraph.com/api/graphql/examples) to fetch file contents without cloning a repository or search for a new framework or API you are using and determine all of the repositories that haven’t been migrated yet. The [Stream API](https://docs.sourcegraph.com/api/stream_api) is for consuming search results as a stream of events, and it can be used to [search](https://docs.sourcegraph.com/api/stream_api#example-curl) over all indexed repositories. Lastly, the [interactive API explorer](https://sourcegraph.com/api/console#%7B%22query%22%3A%22%23%20Type%20queries%20here%2C%20with%20completion%2C%20validation%2C%20and%20hovers.%5Cn%23%5Cn%23%20Here's%20an%20example%20query%20to%20get%20you%20started%3A%5Cn%5Cnquery%20%7B%5Cn%20%20currentUser%20%7B%5Cn%20%20%20%20username%5Cn%20%20%7D%5Cn%20%20repositories(first%3A%201)%20%7B%5Cn%20%20%20%20nodes%20%7B%5Cn%20%20%20%20%20%20name%5Cn%20%20%20%20%7D%5Cn%20%20%7D%5Cn%7D%5Cn%22%7D) is meant for building and testing your API queries. 

Reviewer 2: ```suggestion
GitHub has a REST API for web clients, but it is not yet documented. On the other hand, Sourcegraph offers different APIs that help you access code-related data available on a Sourcegraph instance. The [GraphQL API](https://docs.sourcegraph.com/api/graphql) accesses data stored and computed by Sourcegraph. This API can [fetch](https://docs.sourcegraph.com/api/graphql/examples) file contents without cloning a repository or search for a new API and determine all of the repositories that haven’t migrated to it yet. The [Stream API](https://docs.sourcegraph.com/api/stream_api) supports consuming search results as a stream of events and it can be used to [search](https://docs.sourcegraph.com/api/stream_api#example-curl) over all indexed repositories. Lastly, use the [interactive API explorer](https://sourcegraph.com/api/console#%7B%22query%22%3A%22%23%20Type%20queries%20here%2C%20with%20completion%2C%20validation%2C%20and%20hovers.%5Cn%23%5Cn%23%20Here's%20an%20example%20query%20to%20get%20you%20started%3A%5Cn%5Cnquery%20%7B%5Cn%20%20currentUser%20%7B%5Cn%20%20%20%20username%5Cn%20%20%7D%5Cn%20%20repositories(first%3A%201)%20%7B%5Cn%20%20%20%20nodes%20%7B%5Cn%20%20%20%20%20%20name%5Cn%20%20%20%20%7D%5Cn%20%20%7D%5Cn%7D%5Cn%22%7D) to build and test your API queries. 
```

## Embedded code search and documentation

Sourcegaph’s [Notebooks](https://docs.sourcegraph.com/notebooks) integrate code search with Markdown for knowledge sharing. Notebooks are created with blocks, and each block can be Markdown, a code search query, a live code snippet, a file, or a symbol.

Notebooks pull information directly from your codebase, so the information served in notebooks (via code search blocks, for example) always reflects what is live in your code at that moment in time. By referencing live code, notebooks are useful for onboarding teammates, documenting vulnerabilities, walking through complex parts of a codebase, or keeping track of useful queries. 

GitHub does not currently offer functionality to embed code search within notebooks or documentation.

## Licensing

GitHub is closed source, while Sourcegraph’s code is publicly available. 

Sourcegraph offers two versions of its product: Sourcegraph OSS (just universal code search) and Sourcegraph Enterprise (full code intelligence platform). 

Sourceraph OSS is Apache 2 licensed and allows any developer to use, modify, and redistribute.

> ## Which code search is best for you?

Reviewer 2: This seems like the TLDR that should go at the very top of the doc
